server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -146,7 +255,133 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+        LOGGER.trace(
+            "[{}]: Obtained decision: [{}/{}]",
+            shardRouting.shardId(),
+            allocateDecision.isDecisionTaken(),
+            allocateDecision.getAllocationDecision()
+        );
+        if (allocateDecision.isDecisionTaken() && AllocationDecision.NO == allocateDecision.getAllocationDecision()) {
+            if (LOGGER.isTraceEnabled()) {
+                LOGGER.trace(
+                    "[{}]: Working with decisions: [{}]",
+                    shardRouting.shardId(),
+                    allocateDecision.getNodeDecisions()
+                        .stream()
+                        .map(
+                            n -> n.getCanAllocateDecision()
+                                .getDecisions()
+                                .stream()
+                                .map(d -> d.label() + ": " + d.type())
+                                .collect(Collectors.toList())
+                        )
+                        .collect(Collectors.toList())
+                );
+            }
+            List<NodeAllocationResult> nodeAllocationResults = allocateDecision.getNodeDecisions();
+            boolean diagnosisFound = checkIsAllocationDisabled(diagnosisOutput, shardRouting, nodeAllocationResults);
+            IndexMetadata index = state.metadata().index(shardRouting.index());
+            if (index != null) {
+                diagnosisFound |= checkDataTierRelatedIssues(diagnosisOutput, index, shardRouting, nodeAllocationResults);
+            }
+            if (diagnosisFound == false) {
+                diagnosisOutput.accept(ACTION_CHECK_ALLOCATION_EXPLAIN_API, shardRouting);
+            }
+        }
+    }
+
+    private static Predicate<NodeAllocationResult> nodeHasDeciderResult(String deciderName, Decision.Type outcome) {
+        return (nodeResult) -> nodeResult.getCanAllocateDecision()
+            .getDecisions()
+            .stream()
+            .anyMatch(decision -> deciderName.equals(decision.label()) && outcome == decision.type());
+    }
+
+    private boolean checkIsAllocationDisabled(
+        BiConsumer<UserAction.Definition, ShardRouting> diagnosisOutput,
+        ShardRouting shardRouting,
+        List<NodeAllocationResult> nodeAllocationResults
+    ) {
+        if (nodeAllocationResults.stream().allMatch(nodeHasDeciderResult(EnableAllocationDecider.NAME, Decision.Type.NO))) {
+            diagnosisOutput.accept(ACTION_ENABLE_ALLOCATIONS, shardRouting);
+            return true;
+        }
+        return false;
+    }
+
+    private boolean checkDataTierRelatedIssues(
+        BiConsumer<UserAction.Definition, ShardRouting> diagnosisOutput,
+        IndexMetadata indexMetadata,
+        ShardRouting shardRouting,
+        List<NodeAllocationResult> nodeAllocationResults
+    ) {
+        boolean diagnosisFound = false;
+        if (indexMetadata.getTierPreference().size() > 0) {
+            List<NodeAllocationResult> dataTierNodes = nodeAllocationResults.stream()
```

Comment:
This was a bit confusing to me as for `dataTierNodes` I would've expected a `List<DiscoveryNode>`. Would this be better named `dataTierAllocationResults` or something like that?